include/glow/Optimizer/GraphOptimizer/NodeSplitting.h

```diff
@@ -0,0 +1,287 @@
+      : SplitNodeOption(splitDims), chunkSizes_(chunkSizes) {
+    CHECK_EQ(splitDims.size(), chunkSizes.size())
+        << "Mistmatch between 'splitDims' and 'chunkSizes' array sizes!";
+  }
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const override;
+};
+
+/// Split option used for splitting a node along the dimensions \ref splitDims
+/// with the given chunk weights \ref chunkWeights for each dimension such that
+/// the resulting chunk sizes after splitting will be proportional to the
+/// chunk weights. The chunk weights must be strictly positive and are NOT
+/// required to be normalized. The number of weights (number of chunks) for
+/// each dimension must be lower than the respective dimension size, otherwise
+/// error is thrown. For example when splitting a 1D tensor with size 10 along
+/// the dimension 0 using the weights {1, 4} we obtain two slices with sizes
+/// {2, 8}.
+class SplitNodeByChunkWeights : public SplitNodeOption {
+
+  /// Array of chunk weights for each split dimension.
+  std::vector<std::vector<float>> chunkWeights_;
+
+public:
+  /// Ctor.
+  SplitNodeByChunkWeights(llvm::ArrayRef<size_t> splitDims,
+                          llvm::ArrayRef<std::vector<float>> chunkWeights)
+      : SplitNodeOption(splitDims), chunkWeights_(chunkWeights) {
+    CHECK_EQ(splitDims.size(), chunkWeights.size())
+        << "Mistmatch between 'splitDims' and 'chunkWeights' array sizes!";
+  }
+
+  /// Split a given dimension size \p dimSize along the dimension \p dim.
+  /// \returns the chunks sizes after splitting.
+  std::vector<dim_t> splitAlongDim(size_t dim, dim_t dimSize) const override;
+};
+
+///===---------------------------------------------------------------------===//
+///                             SplitNodeConstraint
+///===---------------------------------------------------------------------===//
+/// Context provided by the node splitting procedure for the user to define node
+/// splitting constraints.
+struct SplitNodeContext {
+  /// Original node which was used for splitting. When this context is provided
+  /// the original node is still part of the graph so the user can also check
+  /// graph connection properties (like whether the original node is followed or
+  /// or preceded by another node, etc).
+  const Node *origNode;
+
+  /// Split nodes obtained by splitting the original node. When this context is
+  /// provided, the split nodes are not yet part of the graph (orphans) and the
+  /// nodes have only temporary SliceNodes attached to their inputs without any
+  /// nodes attached to their outputs so you are only allowed to check the split
+  /// nodes types, attributes, input/output operand types and also the offsets
+  /// of the SliceNodes used for splitting the input operands.
+  std::vector<Node *> splitNodes;
```

Comment:
I used a struct since I thought maybe we would want to add more information in the future. But I followed your advice and I removed the struct entirely and used instead a `const Node *origNode` and a `const std::vector<Node *> &splitNodes` directly in the constraint API.